crates/biome_js_type_info/src/type_info.rs

```diff
@@ -0,0 +1,375 @@
+//! JavaScript files as well.
+//!
+//! This can be used by type-informed lint rules such as `noFloatingPromises`.
+//!
+//! The type information is instantiated and updated inside the Workspace
+//! Server.
+
+use std::str::FromStr;
+
+use biome_rowan::Text;
+use static_assertions::assert_eq_size;
+
+/// Represents an inferred TypeScript type.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub enum Type {
+    /// The type is unknown because inference couldn't determine a type.
+    ///
+    /// This is different from `UnknownKeyword`, because an explicit `unknown`
+    /// should not be counted as a failure of the inference.
+    #[default]
+    Unknown,
+
+    // Primitives
+    BigInt,
+    Boolean,
+    Null,
+    Number,
+    String,
+    Symbol,
+    Undefined,
+
+    // Complex types
+    Array(Box<Type>),
+    Class(Box<Class>),
+    Constructor(Box<Constructor>),
+    Function(Box<Function>),
+    Namespace(Box<Namespace>),
+    Object(Box<Object>),
+    Promise(Box<Type>),
+    Tuple(Box<Tuple>),
+
+    // Compound types
+    Intersection(Box<Intersection>),
+    Union(Box<Union>),
+
+    /// Type derived from another through a built-in operator.
+    TypeOperator(Box<TypeOperatorType>),
+
+    /// Alias to another type.
+    Alias(Box<Type>),
+
+    /// Literal value used as a type.
+    Literal(Box<Literal>),
+
+    /// Reference to another type.
+    Reference(Box<TypeReference>),
+
+    /// Reference to another type through the `typeof` operator.
+    TypeofType(Box<TypeReference>),
+
+    /// The `any` keyword.
+    ///
+    /// This variant may also be used if the `any` keyword is implied.
+    /// For instance, in `catch (e) {}`, `e` is implied to be `any` or
+    /// `unknown`, depending on the TypeScript configuration.
+    AnyKeyword,
+
+    /// The `never` keyword.
+    ///
+    /// Note that unlike TypeScript itself, we never use `never` as a fallback
+    /// for when inference determines that a type can have no values. Instead,
+    /// we will infer [`Type::Unknown`] in such a case, erring on the side of
+    /// caution that the lack of possible values may be a failure of our own
+    /// inference.
+    NeverKeyword,
+
+    /// The `object` keyword.
+    ObjectKeyword,
+
+    /// The `this` keyword.
+    ThisKeyword,
+
+    /// The `unknown` keyword.
+    ///
+    /// This variant may also be used if the `unknown` keyword is implied.
+    /// For instance, in `catch (e) {}`, `e` is implied to be `unknown` or
+    /// `any`, depending on the TypeScript configuration.
+    UnknownKeyword,
+
+    /// The `void` keyword.
+    VoidKeyword,
+}
+
+// `Type` should not be bigger than 16 bytes.
+assert_eq_size!(Type, [usize; 2]);
+
+/// A class definition.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Class {
+    /// Name of the class, if specified in the definition.
+    pub name: Option<Text>,
+
+    /// Class members.
+    pub members: Vec<ClassMember>,
```

Comment:
If it is possible we could reduce the size of the types by using `Box<[]>` instead of `Vec`.

```suggestion
    pub members: Box<[ClassMember]>,
```

Same in other places.

Reply:
Done!